cli/cmd/edges.go

```diff
@@ -0,0 +1,325 @@
+	return cmd
+}
+
+func validateEdgesOutputFormat(options *edgesOptions) error {
+	switch options.outputFormat {
+	case tableOutput, jsonOutput:
+		return nil
+	default:
+		return fmt.Errorf("--output currently only supports %s and %s", tableOutput, jsonOutput)
+	}
+}
+
+func buildEdgesRequests(resources []string, options *edgesOptions) ([]*pb.EdgesRequest, error) {
+	targets, err := util.BuildResources(options.namespace, resources)
+
+	if err != nil {
+		return nil, err
+	}
+	err = validateEdgesOutputFormat(options)
+	if err != nil {
+		return nil, err
+	}
+
+	requests := make([]*pb.EdgesRequest, 0)
+	for _, target := range targets {
+		if target.Name != "" {
+			target.Name = "" // validating that no resource name is specified in the request
+		}
+
+		requestParams := util.EdgesRequestParams{
+			ResourceType: target.Type,
+			Namespace:    options.namespace,
+		}
+
+		req, err := util.BuildEdgesRequest(requestParams)
+		if err != nil {
+			return nil, err
+		}
+		requests = append(requests, req)
+	}
+	return requests, nil
+}
+
+func edgesRespToRows(resp *pb.EdgesResponse) []*pb.Edge {
+	rows := make([]*pb.Edge, 0)
+	if resp != nil {
+		for _, edgeTable := range resp.GetOk().Edges {
+			rows = append(rows, edgeTable)
+		}
+	}
+	return rows
+}
+
+func requestEdgesFromAPI(client pb.ApiClient, req *pb.EdgesRequest) (*pb.EdgesResponse, error) {
+	resp, err := client.Edges(context.Background(), req)
+	if err != nil {
+		return nil, fmt.Errorf("Edges API error: %+v", err)
+	}
+	if e := resp.GetError(); e != nil {
+		return nil, fmt.Errorf("Edges API response error: %+v", e.Error)
+	}
+	return resp, nil
+}
+
+func renderEdgeStats(rows []*pb.Edge, options *edgesOptions) string {
+	var buffer bytes.Buffer
+	w := tabwriter.NewWriter(&buffer, 0, 0, padding, ' ', tabwriter.AlignRight)
+	writeEdgesToBuffer(rows, w, options)
+	w.Flush()
+
+	return renderEdges(buffer, options)
+}
+
+type edgeRowStats struct {
+	src    string
+	dst    string
+	client string
+	server string
+	msg    string
+}
+
+func writeEdgesToBuffer(rows []*pb.Edge, w *tabwriter.Writer, options *edgesOptions) {
+	edgeTables := make(map[string]*edgeRowStats)
+	if len(rows) != 0 {
+		for _, r := range rows {
+			key := string(r.Dst.Name + r.Src.Name)
+			clientID := r.ClientId
+			serverID := r.ServerId
+			if len(clientID) > 0 {
+				parts := strings.Split(clientID, ".")
+				clientID = parts[0] + "." + parts[1]
+			}
+			if len(serverID) > 0 {
+				parts := strings.Split(serverID, ".")
+				serverID = parts[0] + "." + parts[1]
+			}
+
+			edgeTables[key] = &edgeRowStats{
+				client: clientID,
+				server: serverID,
+				msg:    r.NoIdentityMsg,
+				src:    r.Src.Name,
+				dst:    r.Dst.Name,
+			}
+		}
+	}
+	switch options.outputFormat {
+	case tableOutput:
+		if len(edgeTables) == 0 {
+			fmt.Fprintln(os.Stderr, "No edges found.")
+			os.Exit(0)
+		}
+		printSingleEdgeTable(edgeTables, w)
```

Comment:
just curious: why is this method called `printSingleEdgeTable`? is it because we're showing one side of a potentially bidirectional edge?

Reply:
Great question 👁. In the future you should be able to run `linkerd edges` for multiple resource types and `printSingleEdgeTable` will be called each time, a la `stat.go`, but this specific PR doesn't support that yet. For clarity I'll change it to `printEdgeTable`.